### PR TITLE
Added JWT.Extensions.DependencyInjection 

### LINKED ISSUE #336: Add NuGet package for integration with Microsoft.Extensions.DependencyInjection
It would be useful and convenient to offer a set of extension methods that would register in the DI container all necessary classes, e.g.:

- `AddJwtEncoder()`
- `AddJwtDecoder()`

### PR DESCRIPTION
**Pull request by abatishchev**
Resolves #336

### CODE REVIEW

**Comment by abatishchev** [comment]
@frankely can you please also help testing [version 7.4.0](https://dev.azure.com/abatishchev/OpenSource/_build/results?buildId=832&view=artifacts&pathAsName=false&type=publishedArtifacts)?

the difference is that `7.3.1` contains all extension methods but `7.4.0` depends on the new `Jwt.Extensions.DependencyInjection` which contains generic DI container extension methods and `Jwt.Extensions.AspNetCore` contains only for authentication.

Makes sense?

**Comment by frankely** [comment]
You bet I got you @abatishchev

**Comment by abatishchev** [comment]
@frankely gentle ping, please. your help is greatly appreciated.

**Comment by frankely** [comment]
I'll get into it! Somehow I had troubles trying to find the new version. I'll try again today!

**Review by frankely — APPROVED** [review]

**Review by frankely — APPROVED** [review]

File: src/JWT.Extensions.AspNetCore/AuthenticationBuilderExtensions.cs [hunk 1]
```diff
@@ -45,7 +40,7 @@ public static AuthenticationBuilder AddJwt<TFactory>(this AuthenticationBuilder
         public static AuthenticationBuilder AddJwt<TFactory>(this AuthenticationBuilder builder, string authenticationScheme, Action<JwtAuthenticationOptions> configureOptions)
             where TFactory : class, IAlgorithmFactory
         {
-            builder.Services.TryAddSingleton<IAlgorithmFactory, TFactory>();
+            builder.Services.AddJwtDecoder<TFactory>();
```
Comment (frankely) [inline src/JWT.Extensions.AspNetCore/AuthenticationBuilderExtensions.cs:43]: This is awesome! We won't have to add the JwtDecoder manually anymore!

**Review by frankely — APPROVED** [review]

**Comment by frankely** [comment]
Done @abatishchev

**Comment by abatishchev** [comment]
You got it [from here](https://abatishchev.visualstudio.com/OpenSource/_build/results?buildId=847&view=artifacts&pathAsName=false&type=publishedArtifacts), correct? Looks good? Ship it?

**Comment by frankely** [comment]
Actually, after upgrading them I'm getting errors because there is no AddJwt extension method anymore and the JwtAuthenticationDefaults and JwtAuthenticationOptions classes don't exist anymore. @abatishchev

**Comment by abatishchev** [comment]
@frankely wait, how come? can you post a snippet of your code? i haven't changed the extension method's signature:

```c#
namespace JWT
{
    public static class AuthenticationBuilderExtensions
    {
        public static AuthenticationBuilder AddJwt(this AuthenticationBuilder builder) { ... }

        public static AuthenticationBuilder AddJwt<TFactory>(this AuthenticationBuilder builder)
            where TFactory : class, IAlgorithmFactory { ... }
```

**Comment by frankely** [comment]
Sorry that I haven't looked into this, but I have been swamped with other stuff. Can you provide an example of how to use the newer version? Do you have available as a NuGet prerelease version somewhere with some docs?

**Comment by abatishchev** [comment]
@frankely sorry for the delay in response! and thanks for helping me to test the new functionality.

I think the easiest way to test would be to download the packages into a folder and setup is as a [local feed](https://docs.microsoft.com/en-us/nuget/hosting-packages/local-feeds), if you haven't done that already.

However the problem is that the package version hasn't changed so the previous version has been cached by NuGet in few places locally. So what I did is I appended both packages with `-beta1 and will increment it every time I make a change in the PR. Will drop it once we're done.

Please try out the latest version [from this build](https://dev.azure.com/abatishchev/OpenSource/_build/results?buildId=868&view=artifacts&pathAsName=false&type=publishedArtifacts).

**Comment by abatishchev** [comment]
@frankely, ping please.

**Comment by frankely** [comment]
Hey! I'll look into this!

**Comment by frankely** [comment]
I upgraded to  JWT to Version="8.4.2" and it works great!

**Review by frankely — APPROVED** [review]

**Review by frankely — APPROVED** [review]

**Review by frankely — APPROVED** [review]

**Comment by abatishchev** [comment]
Pushed to NuGet as:

- https://www.nuget.org/packages/JWT/8.6.0
- https://www.nuget.org/packages/JWT.Extensions.AspNetCore/8.1.0
- https://www.nuget.org/packages/JWT.Extensions.DependencyInjection/1.0.0

**Pull request merged**